packages/components/src/internal/components/user/UserLink.tsx

```diff
@@ -1,30 +1,90 @@
 
-interface Props {
-    allUsers: List<IUser>;
+import { caseInsensitive } from '../../util/utils';
+
+import { selectRowsUserProps, UserDetailsPanel } from './UserDetailsPanel';
+
+interface UserLinkProps {
     currentUser: User;
-    userId: string;
+    userDisplayValue?: string;
+    userId: number;
 }
 
-export class UserLink extends React.PureComponent<Props> {
-    render() {
-        const { allUsers, currentUser, userId } = this.props;
-        if (!allUsers || !userId) return null;
-
-        let targetUser: IUser = null;
-        if (allUsers) {
-            targetUser = allUsers.find(user => user.userId === parseInt(userId));
-        }
-
-        if (targetUser) {
-            const link = AppURL.create('q', 'core', 'siteusers', userId).toHref();
-            return currentUser.isAdmin ? <a href={link}>{targetUser.displayName}</a> : targetUser.displayName;
-        }
-        return null;
+export const UserLink: FC<UserLinkProps> = props => {
+    const { currentUser, userId, userDisplayValue } = props;
+    const [showDetails, setShowDetails] = useState<boolean>(false);
+    const [targetUserDisplayValue, setTargetUserDisplayValue] = useState<string>();
+    const isSelf = userId === currentUser.id;
+
+    useEffect(() => {
+        (async () => {
+            try {
+                if (!!userId && userDisplayValue === undefined) {
+                    const targetUser = await selectRowsUserProps(userId);
+                    setTargetUserDisplayValue(caseInsensitive(targetUser, 'DisplayName'));
+                } else {
+                    setTargetUserDisplayValue(userDisplayValue);
+                }
+            } catch (error) {
+                console.error(error);
+            }
+        })();
+    }, [userDisplayValue, userId]);
+
+    const toggleDetailsModal = useCallback(() => {
+        setShowDetails(current => !current);
+    }, []);
+
+    if (!userId) return null;
+
+    if (!isSelf && (!userCanReadUserDetails(currentUser) || !targetUserDisplayValue)) {
+        return <span>{targetUserDisplayValue ?? userId}</span>;
     }
+
+    return (
+        <>
+            <a onClick={toggleDetailsModal} style={{ cursor: 'pointer' }}>
+                {targetUserDisplayValue}
+            </a>
+            {showDetails && (
+                <UserDetailsPanel
+                    currentUser={currentUser}
+                    userId={userId}
+                    toggleDetailsModal={toggleDetailsModal}
+                    isSelf={!userCanReadUserDetails(currentUser) && isSelf}
+                />
+            )}
+        </>
+    );
+};
+
+interface UserOrGroup {
+    displayName: string;
+    id: number;
+    type?: string;
+}
+
+interface UserLinkListProps {
+    currentUser: User;
+    users: UserOrGroup[];
 }
+
+export const UserLinkList: FC<UserLinkListProps> = ({ currentUser, users }) => {
+    return (
+        <>
+            {users.map((u, i) => (
+                <>
+                    {i > 0 && <>, </>}
```

Comment:
I guess this is doing the right thing, but I'm not parsing it well.

Reply:
Note sure why I was using a fragment here. Updated to be `{i > 0 && ', '}`